src/caffe/util/db.cpp

```diff
@@ -0,0 +1,84 @@
+#include "caffe/util/db.hpp"
+
+#include <sys/stat.h>
+#include <string>
+
+namespace caffe { namespace db {
+
+const size_t LMDB_MAP_SIZE = 1099511627776;  // 1 TB
+
+void LevelDB::Open(const string& source, Mode mode) {
+  leveldb::Options options;
+  options.block_size = 65536;
+  options.write_buffer_size = 268435456;
+  options.max_open_files = 100;
+  options.error_if_exists = mode == NEW;
+  options.create_if_missing = mode != READ;
+  leveldb::Status status = leveldb::DB::Open(options, source, &db_);
+  CHECK(status.ok()) << "Failed to open leveldb " << source
+                     << std::endl << status.ToString();
+  LOG(INFO) << "Opened leveldb " << source;
+}
+
+void LMDB::Open(const string& source, Mode mode) {
+  MDB_CHECK(mdb_env_create(&mdb_env_));
+  MDB_CHECK(mdb_env_set_mapsize(mdb_env_, LMDB_MAP_SIZE));
+  if (mode == NEW) {
+    CHECK_EQ(mkdir(source.c_str(), 0744), 0) << "mkdir " << source << "failed";
+  }
+  int flags = 0;
+  if (mode == READ) {
+    flags = MDB_RDONLY | MDB_NOTLS;
+  }
+  MDB_CHECK(mdb_env_open(mdb_env_, source.c_str(), flags, 0664));
+  LOG(INFO) << "Opened lmdb " << source;
+}
+
+LMDBCursor* LMDB::NewCursor() {
+  MDB_txn* mdb_txn;
+  MDB_cursor* mdb_cursor;
+  MDB_CHECK(mdb_txn_begin(mdb_env_, NULL, MDB_RDONLY, &mdb_txn));
```

Comment:
If for every cursor one creates a new mdb_txn, it can run out of readers pretty quickly.